File: android/shared/src/main/java/com/horizend/crmdelivery/shared/api/CustomerApiClient.kt
package com.horizend.crmdelivery.shared.api

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.header
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.patch
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json
import java.util.UUID

object CustomerApiClient {
    private const val customerApiBaseUrl = "https://crm.ayartuerk.me/api/v1"

    private val httpClient = HttpClient {
        expectSuccess = true
        install(ContentNegotiation) {
            json(
                Json {
                    encodeDefaults = true
                    ignoreUnknownKeys = true
                    isLenient = true
                }
            )
        }
    }

    private fun bearer(accessToken: String): String = "Bearer $accessToken"

    private fun idempotencyKey(): String = UUID.randomUUID().toString().replace("-", "")

    suspend fun startCustomerEmailAuthentication(
        email: String,
        language: String,
        initiationNonce: String
    ): CustomerEmailAuthStartResponse =
        httpClient.post("$customerApiBaseUrl/customer/auth/email/start") {
            header("Idempotency-Key", idempotencyKey())
            contentType(ContentType.Application.Json)
            setBody(
                CustomerEmailAuthStartRequest(
                    email = email,
                    locale = language,
                    initiationNonce = initiationNonce
                )
            )
        }.body()

    suspend fun completeCustomerEmailAuthentication(
        attemptId: String,
        manualCode: String,
        initiationNonce: String
    ): CustomerEmailAuthCompleteResponse =
        httpClient.post("$customerApiBaseUrl/customer/auth/email/complete") {
            header("Idempotency-Key", idempotencyKey())
            contentType(ContentType.Application.Json)
            setBody(
                CustomerEmailAuthCompleteRequest(
                    attemptId = attemptId,
                    manualCode = manualCode,
                    initiationNonce = initiationNonce
                )
            )
        }.body()

    suspend fun getCustomerEmailAuthenticationSession(
        accessToken: String
    ): CustomerEmailAuthSessionResponse =
        httpClient.get("$customerApiBaseUrl/customer/auth/session") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun logoutCustomerEmailAuthentication(
        accessToken: String
    ): CustomerLogoutResponse =
        httpClient.post("$customerApiBaseUrl/customer/auth/logout") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun getCustomerProducts(): List<CustomerProduct> =
        httpClient.get("$customerApiBaseUrl/public/catalog")
            .body<CustomerCatalogResponse>()
            .catalog
            .products

    suspend fun startCustomerSession(
        deviceId: String,
        platform: String,
        appVersion: String,
        fullName: String,
        username: String = "",
        language: String = "en"
    ): CustomerSessionStartResponse =
        httpClient.post("$customerApiBaseUrl/customer/session/start") {
            contentType(ContentType.Application.Json)
            setBody(
                CustomerSessionStartRequest(
                    deviceId = deviceId,
                    platform = platform,
                    appVersion = appVersion,
                    fullName = fullName,
                    username = username,
                    language = language
                )
            )
        }.body()

    suspend fun verifyCustomerSession(accessToken: String): CustomerSessionVerifyResponse =
        httpClient.post("$customerApiBaseUrl/customer/session/verify") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun logoutCustomerSession(accessToken: String): CustomerLogoutResponse =
        httpClient.post("$customerApiBaseUrl/customer/session/logout") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun getCustomerProfile(accessToken: String): CustomerProfileResponse =
        httpClient.get("$customerApiBaseUrl/customer/me") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun updateCustomerProfile(
        accessToken: String,
        fullName: String,
        username: String = "",
        preferredLanguage: String = "en"
    ): CustomerProfileResponse =
        httpClient.patch("$customerApiBaseUrl/customer/me") {
            header("Authorization", bearer(accessToken))
            contentType(ContentType.Application.Json)
            setBody(
                CustomerProfileUpdateRequest(
                    fullName = fullName,
                    username = username,
                    preferredLanguage = preferredLanguage
                )
            )
        }.body()

    suspend fun getCustomerLocations(accessToken: String): CustomerLocationsResponse =
        httpClient.get("$customerApiBaseUrl/customer/locations") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun createCustomerLocation(
        accessToken: String,
        label: String = "",
        address: String = "",
        googleMapsLink: String = "",
        latitude: String = "",
        longitude: String = "",
        saveAsPreferred: Boolean = false
    ): CustomerLocationResponse =
        httpClient.post("$customerApiBaseUrl/customer/locations") {
            header("Authorization", bearer(accessToken))
            contentType(ContentType.Application.Json)
            setBody(
                CreateCustomerLocationRequest(
                    label = label,
                    address = address,
                    googleMapsLink = googleMapsLink,
                    latitude = latitude,
                    longitude = longitude,
                    saveAsPreferred = saveAsPreferred
                )
            )
        }.body()

    suspend fun getCustomerCart(accessToken: String): CustomerCartResponse =
        httpClient.get("$customerApiBaseUrl/customer/cart") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun addCustomerCartItem(accessToken: String, productId: Int, quantity: Int): CustomerCartResponse =
        httpClient.post("$customerApiBaseUrl/customer/cart/items") {
            header("Authorization", bearer(accessToken))
            contentType(ContentType.Application.Json)
            setBody(AddCartItemRequest(productId = productId, quantity = quantity))
        }.body()

    suspend fun updateCustomerCartItem(accessToken: String, itemId: Int, quantity: Int): CustomerCartResponse =
        httpClient.patch("$customerApiBaseUrl/customer/cart/items/$itemId") {
            header("Authorization", bearer(accessToken))
            contentType(ContentType.Application.Json)
            setBody(UpdateCartItemRequest(quantity = quantity))
        }.body()

    suspend fun removeCustomerCartItem(accessToken: String, itemId: Int): CustomerCartResponse =
        httpClient.delete("$customerApiBaseUrl/customer/cart/items/$itemId") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun checkoutCustomerCart(
        accessToken: String,
        deliveryAddress: String = "",
        notes: String,
        savedLocationId: Int? = null,
        usePreferredLocation: Boolean = false,
        locationLabel: String = deliveryAddress,
        googleMapsLink: String = "",
        latitude: String = "",
        longitude: String = "",
        saveAsPreferred: Boolean = false
    ): CustomerOrderResponse =
        httpClient.post("$customerApiBaseUrl/customer/checkout/address") {
            header("Authorization", bearer(accessToken))
            contentType(ContentType.Application.Json)
            setBody(
                CheckoutRequest(
                    savedLocationId = savedLocationId,
                    usePreferredLocation = usePreferredLocation,
                    address = deliveryAddress,
                    locationLabel = locationLabel,
                    googleMapsLink = googleMapsLink,
                    latitude = latitude,
                    longitude = longitude,
                    deliveryNote = notes,
                    saveAsPreferred = saveAsPreferred
                )
            )
        }.body()

    suspend fun checkoutCustomerPickup(
        accessToken: String,
        notes: String = "",
        paymentMethodCode: String = ""
    ): CustomerOrderResponse =
        httpClient.post("$customerApiBaseUrl/customer/checkout/pickup") {
            header("Authorization", bearer(accessToken))
            contentType(ContentType.Application.Json)
            setBody(
                CheckoutPickupRequest(
                    notes = notes,
                    paymentMethodCode = paymentMethodCode
                )
            )
        }.body()

    suspend fun getCustomerOrders(accessToken: String): CustomerOrdersResponse =
        httpClient.get("$customerApiBaseUrl/customer/orders") {
            header("Authorization", bearer(accessToken))
        }.body()

    suspend fun getCustomerOrderDetail(accessToken: String, orderId: Int): CustomerOrderResponse =
        httpClient.get("$customerApiBaseUrl/customer/orders/$orderId") {
            header("Authorization", bearer(accessToken))
        }.body()
}

File: apple/shared/Sources/CustomerApiClient.swift
import Foundation

public enum CustomerApiClient {
    private static let customerApiBaseURL = "https://crm.ayartuerk.me/api/v1"

    public static func getCustomerProducts() async throws -> [CustomerProduct] {
        let response: CustomerCatalogResponse = try await fetch(URL(string: "\(customerApiBaseURL)/public/catalog")!)
        return response.catalog.products
    }

    public static func startCustomerSession(
        deviceId: String,
        platform: String,
        appVersion: String,
        fullName: String,
        username: String = "",
        language: String = "en"
    ) async throws -> CustomerSessionStartResponse {
        let request = CustomerSessionStartRequest(
            deviceId: deviceId,
            platform: platform,
            appVersion: appVersion,
            fullName: fullName,
            username: username,
            language: language
        )
        return try await post(URL(string: "\(customerApiBaseURL)/customer/session/start")!, body: request)
    }


    public static func verifyCustomerSession(accessToken: String) async throws -> CustomerSessionVerifyResponse {
        try await post(URL(string: "\(customerApiBaseURL)/customer/session/verify")!, body: EmptyRequest(), accessToken: accessToken)
    }

    public static func logoutCustomerSession(accessToken: String) async throws -> CustomerLogoutResponse {
        try await post(URL(string: "\(customerApiBaseURL)/customer/session/logout")!, body: EmptyRequest(), accessToken: accessToken)
    }

    public static func getCustomerProfile(accessToken: String) async throws -> CustomerProfileResponse {
        try await fetch(URL(string: "\(customerApiBaseURL)/customer/me")!, accessToken: accessToken)
    }

    public static func updateCustomerProfile(
        accessToken: String,
        fullName: String,
        username: String = "",
        preferredLanguage: String = "en"
    ) async throws -> CustomerProfileResponse {
        let request = CustomerProfileUpdateRequest(
            fullName: fullName,
            username: username,
            preferredLanguage: preferredLanguage
        )
        return try await patch(URL(string: "\(customerApiBaseURL)/customer/me")!, body: request, accessToken: accessToken)
    }

    public static func getCustomerLocations(accessToken: String) async throws -> CustomerLocationsResponse {
        try await fetch(URL(string: "\(customerApiBaseURL)/customer/locations")!, accessToken: accessToken)
    }

    public static func createCustomerLocation(
        accessToken: String,
        label: String = "",
        address: String = "",
        googleMapsLink: String = "",
        latitude: String = "",
        longitude: String = "",
        saveAsPreferred: Bool = false
    ) async throws -> CustomerLocationResponse {
        let request = CreateCustomerLocationRequest(
            label: label,
            address: address,
            googleMapsLink: googleMapsLink,
            latitude: latitude,
            longitude: longitude,
            saveAsPreferred: saveAsPreferred
        )

        return try await post(URL(string: "\(customerApiBaseURL)/customer/locations")!, body: request, accessToken: accessToken)
    }

    public static func getCustomerCart(accessToken: String) async throws -> CustomerCartResponse {
        try await fetch(URL(string: "\(customerApiBaseURL)/customer/cart")!, accessToken: accessToken)
    }

    public static func addCustomerCartItem(accessToken: String, productId: Int, quantity: Int) async throws -> CustomerCartResponse {
        let request = AddCartItemRequest(productId: productId, quantity: quantity)
        return try await post(URL(string: "\(customerApiBaseURL)/customer/cart/items")!, body: request, accessToken: accessToken)
    }

    public static func updateCustomerCartItem(accessToken: String, itemId: Int, quantity: Int) async throws -> CustomerCartResponse {
        let request = UpdateCartItemRequest(quantity: quantity)
        return try await patch(URL(string: "\(customerApiBaseURL)/customer/cart/items/\(itemId)")!, body: request, accessToken: accessToken)
    }

    public static func removeCustomerCartItem(accessToken: String, itemId: Int) async throws -> CustomerCartResponse {
        try await delete(URL(string: "\(customerApiBaseURL)/customer/cart/items/\(itemId)")!, accessToken: accessToken)
    }

    public static func checkoutCustomerCart(
        accessToken: String,
        deliveryAddress: String = "",
        notes: String,
        savedLocationId: Int? = nil,
        usePreferredLocation: Bool = false,
        locationLabel: String? = nil,
        googleMapsLink: String = "",
        latitude: String = "",
        longitude: String = "",
        saveAsPreferred: Bool = false
    ) async throws -> CustomerOrderResponse {
        let request = CheckoutRequest(
            savedLocationId: savedLocationId,
            usePreferredLocation: usePreferredLocation,
            address: deliveryAddress,
            locationLabel: locationLabel ?? deliveryAddress,
            googleMapsLink: googleMapsLink,
            latitude: latitude,
            longitude: longitude,
            deliveryNote: notes,
            saveAsPreferred: saveAsPreferred
        )
        return try await post(URL(string: "\(customerApiBaseURL)/customer/checkout/address")!, body: request, accessToken: accessToken)
    }

    public static func checkoutCustomerPickup(
        accessToken: String,
        notes: String = "",
        paymentMethodCode: String = ""
    ) async throws -> CustomerOrderResponse {
        let request = CheckoutPickupRequest(notes: notes, paymentMethodCode: paymentMethodCode)
        return try await post(URL(string: "\(customerApiBaseURL)/customer/checkout/pickup")!, body: request, accessToken: accessToken)
    }

    public static func getCustomerOrders(accessToken: String) async throws -> CustomerOrdersResponse {
        try await fetch(URL(string: "\(customerApiBaseURL)/customer/orders")!, accessToken: accessToken)
    }

    public static func getCustomerOrderDetail(accessToken: String, orderId: Int) async throws -> CustomerOrderResponse {
        try await fetch(URL(string: "\(customerApiBaseURL)/customer/orders/\(orderId)")!, accessToken: accessToken)
    }

    private static func fetch<ResponseBody: Decodable>(_ url: URL, accessToken: String? = nil) async throws -> ResponseBody {
        var request = URLRequest(url: url)
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "authorization")
        }

        let (data, response) = try await URLSession.shared.data(for: request)

        if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
            throw URLError(.badServerResponse)
        }

        return try JSONDecoder().decode(ResponseBody.self, from: data)
    }

    private static func patch<RequestBody: Encodable, ResponseBody: Decodable>(_ url: URL, body: RequestBody, accessToken: String? = nil) async throws -> ResponseBody {
        var request = URLRequest(url: url)
        request.httpMethod = "PATCH"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "authorization")
        }
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)

        if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
            throw URLError(.badServerResponse)
        }

        return try JSONDecoder().decode(ResponseBody.self, from: data)
    }

    private static func delete<ResponseBody: Decodable>(_ url: URL, accessToken: String? = nil) async throws -> ResponseBody {
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "authorization")
        }

        let (data, response) = try await URLSession.shared.data(for: request)

        if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
            throw URLError(.badServerResponse)
        }

        return try JSONDecoder().decode(ResponseBody.self, from: data)
    }

    private static func post<RequestBody: Encodable, ResponseBody: Decodable>(_ url: URL, body: RequestBody, accessToken: String? = nil) async throws -> ResponseBody {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "authorization")
        }
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)

        if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
            throw URLError(.badServerResponse)
        }

        return try JSONDecoder().decode(ResponseBody.self, from: data)
    }
}
